### PR TITLE
feat: add rho-Pollard example

### PR DESCRIPTION
**Pull request by vil02**
This PR adds an example of factorization using [Pollard's-rho algorithm](https://en.wikipedia.org/wiki/Pollard%27s_rho_algorithm). This is related to #70.

### CODE REVIEW

File: examples/rho_pollard.nim [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    return some(divisor)
+
+
+func somePoly(number: BigInt): BigInt =
+  return number*number+1.initBigInt
+
+
+proc main() =
+  const someNum = "44077431694086786329".initBigInt
+  let result = rhoPollard(someNum, somePoly)
```
Comment (konsumlamm) [inline examples/rho_pollard.nim:38]:
```suggestion
  let result = pollardRho(someNum, somePolynomial)
```

File: examples/rho_pollard.nim [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  if result.isSome():
+    echo fmt"{result.get()} is a factor of {someNum}"
+    assert someNum mod result.get() == 0.initBigInt
```
Comment (konsumlamm) [inline examples/rho_pollard.nim:41]:
```suggestion
  if result.isSome():
    let factor = result.get()
    echo fmt"{factor} is a factor of {someNum}"
    assert someNum mod factor == 0
```

File: examples/rho_pollard.nim [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  else:
+    echo fmt"could not find a factor of {someNum} using rho-Pollard algorithm"
```
Comment (konsumlamm) [inline examples/rho_pollard.nim:43]:
```suggestion
    echo fmt"could not find a factor of {someNum}"
```
This doesn't mean that one can't find a factor using the algorithm, one might have to use a different polynomial or a different initial value.

File: examples/rho_pollard.nim [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    hare = initialValue
+    divisor = 1.initBigInt
+
+  while divisor == 1.initBigInt:
+    turtle = nextIterationMod(turtle)
+    hare = nextIterationMod(nextIterationMod(hare))
+    divisor = gcd(turtle-hare, n)
+
+  if divisor != n:
+    return some(divisor)
```
Comment (konsumlamm) [inline examples/rho_pollard.nim:29]:
```suggestion
func pollardRho(
  n: BigInt,
  polynomial: proc(x: BigInt): BigInt {.noSideEffect.},
  initialValue: BigInt = 2.initBigInt): Option[BigInt] =
  ## Performs Pollard's rho algorithm to find a non-trivial factor of `n`.
  func polynomialMod(x: BigInt): BigInt =
    polynomial(x) mod n

  var
    turtle = initialValue
    hare = initialValue
    divisor = 1.initBigInt

  while divisor == 1:
    turtle = polynomialMod(turtle)
    hare = polynomialMod(polynomialMod(hare))
    divisor = gcd(turtle - hare, n)

  if divisor != n:
    some(divisor)
  else:
    none(BigInt)
```

File: examples/rho_pollard.nim [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+
+func somePoly(number: BigInt): BigInt =
+  return number*number+1.initBigInt
```
Comment (konsumlamm) [inline examples/rho_pollard.nim:33]:
```suggestion
func somePolynomial(x: BigInt): BigInt =
  x * x + 1.initBigInt
```

File: examples/rho_pollard.nim (hunk 4 above)
Comment (vil02) [reply examples/rho_pollard.nim:29]: Why do you insist on having the _else_ branch and explicitly returning the `none(BigInt)`?

File: examples/rho_pollard.nim [hunk 6]
```diff
@@ -0,0 +1,46 @@
+## Rho-Pollard
+##
+## This file illustrates how to find a factor of an integer using the
+## [rho-Pollard algorithm](https://en.wikipedia.org/wiki/Pollard%27s_rho_algorithm).
```
[the remark this answers is not in the document]
Comment (vil02) [reply examples/rho_pollard.nim]: Renaming is done in d785567d3c42aeba1fc7a2b8a2877ab50494d06b.

**Comment by vil02** [comment]
@konsumlamm thanks for a review!

File: examples/rho_pollard.nim (hunk 4 above)
Comment (konsumlamm) [reply examples/rho_pollard.nim]: I think it's clearer to explicitly return `none(BigInt)` than to rely on it being the default value for `Option`s.

Comment (vil02) [reply examples/rho_pollard.nim]: Please have a look at 145dc8858b6e24a64dc956d222db47e5a71c6a3b.

**Review by konsumlamm — APPROVED** [review]

**Review by dlesnoff — APPROVED** [review]

**Comment by konsumlamm** [comment]
@narimiran can this be merged?

**Pull request merged**